File: 프로그래머스/타겟 넘버.py
from collections import deque

def solution(numbers, target):
    answer = 0
    queue = deque()
    queue.append((numbers[0], 0))
    queue.append((-numbers[0], 0))
    
    while queue:
        temp, idx = queue.popleft()
        if idx == len(numbers)-1:
            if temp == target:
                answer += 1
        else:
            idx += 1
            queue.append((temp + numbers[idx], idx))
            queue.append((temp - numbers[idx], idx))
        
    return answer

File: 프로그래머스/타겟 넘버.java
class Solution {
    private int rec(int[] numbers, int target, int idx, int sum) {
        if (idx == numbers.length-1) {
            if (sum == target)
                return 1;
            return 0;
        }
        return rec(numbers, target, idx+1, sum-numbers[idx+1]) + rec(numbers, target, idx+1, sum+numbers[idx+1]);
    }

    public int solution(int[] numbers, int target) {
        return rec(numbers, target, 0, -numbers[0]) + rec(numbers, target, 0, numbers[0]);
    }
}